### ISSUE TITLE
sccache direct mode doesn't restore gcc dependency files

### ISSUE THREAD

**Issue by AJIOB** [mozilla/sccache#2073]
# Build commands (from CMake verbose mode)

## Without direct mode
```sh
cd /workspaces/custom_project/cmake-build-vscode/src && env SCCACHE_DIR=/workspaces/custom_project/.sccache/docker-vscode /usr/local/bin/sccache /usr/bin/g++ -DBOOST_ALL_NO_LIB -DBOOST_BEAST_NO_POSIX_FADVISE -DBOOST_THREAD_PROVIDES_THREAD_DESTRUCTOR_CALLS_TERMINATE_IF_JOINABLE -DBOOST_THREAD_PROVIDES_THREAD_MOVE_ASSIGN_CALLS_TERMINATE_IF_JOINABLE -I/workspaces/custom_project/src -ffunction-sections -fdata-sections -flto -ffat-lto-objects -Wall -ggdb3 -DNDEBUG  -fno-lto -O0 -UNDEBUG --coverage -fvisibility=hidden -fvisibility-inlines-hidden -std=c++14 -MD -MT src/CMakeFiles/ProjLib.dir/util/version/version_provider.cpp.o -MF CMakeFiles/ProjLib.dir/util/version/version_provider.cpp.o.d -o CMakeFiles/ProjLib.dir/util/version/version_provider.cpp.o -c /workspaces/custom_project/src/util/version/version_provider.cpp
```

## With direct mode
```sh
cd /workspaces/custom_project/cmake-build-vscode/src && env SCCACHE_DIR=/workspaces/custom_project/.sccache/docker-vscode SCCACHE_DIRECT=true /usr/local/bin/sccache /usr/bin/g++ -DBOOST_ALL_NO_LIB -DBOOST_BEAST_NO_POSIX_FADVISE -DBOOST_THREAD_PROVIDES_THREAD_DESTRUCTOR_CALLS_TERMINATE_IF_JOINABLE -DBOOST_THREAD_PROVIDES_THREAD_MOVE_ASSIGN_CALLS_TERMINATE_IF_JOINABLE -I/workspaces/custom_project/src -ffunction-sections -fdata-sections -flto -ffat-lto-objects -Wall -ggdb3 -DNDEBUG  -fno-lto -O0 -UNDEBUG --coverage -fvisibility=hidden -fvisibility-inlines-hidden -std=c++14 -MD -MT src/CMakeFiles/ProjLib.dir/util/version/version_provider.cpp.o -MF CMakeFiles/ProjLib.dir/util/version/version_provider.cpp.o.d -o CMakeFiles/ProjLib.dir/util/version/version_provider.cpp.o -c /workspaces/custom_project/src/util/version/version_provider.cpp
```

# Environment
* OS debian 9 (Linux)
* gcc 6.3.0 (native GCC from OS)
* CMake 3.23.1
* sccache 0.7.6 with local dir cache
* sccache environment variables:
  * SCCACHE_IDLE_TIMEOUT=0
  * SCCACHE_CACHE_SIZE=4096M

# Problem
When I use *SCCACHE_DIRECT=true* environment variable, sccache doesn't restore **.d** file (CMakeFiles/ProjLib.dir/util/version/version_provider.cpp.o.d) while it restores the object file one.

CMake uses **.d** files for incremental builds & if changes were not found, it doesn't trigger required files to rebuild: if I change the header file, the dependent source file (that was restored by sccache itself) is not rebuilt.

Looks like in preprocessor mode (while *SCCACHE_DIRECT=true* is not used) **.d** file is generated by the compiler itself while preprocessing is performed by the compiler.

# Temporary workaround

Don't use *SCCACHE_DIRECT=true*, because it has problems with CMake incremental builds.

But classic preprocessor mode downgrades the speed, as #160 told.

# Wanted solution

Direct sccache mode should restore/regenerate **.d** files too if they are generated in the direct mode.

If I correctly understand, sccache has already used the list of recursive dependent headers for hash generating.

**Comment by AJIOB** [comment]
One more addition:
ccache resolves similar problems with the depend mode, as described [here](https://ccache.dev/manual/4.9.html#_the_depend_mode)

**Comment by sylvestre** [comment]
Interesting. Not sure we will have time to work on it (we don't use Cmake in this context).
Patches welcome!

**Comment by cooljeanius** [comment]
Related to #1032 perhaps?

**Comment by glandium** [comment]
This was fixed with #2195
